pkg/cache/sql/db/client_test.go

```diff
@@ -0,0 +1,515 @@
+			for _, v := range a {
+				vk := v.(*string)
+				*vk = string(toBytes(testObject.Id))
+			}
+		})
+		r.EXPECT().Err().Return(nil)
+		r.EXPECT().Next().Return(false)
+		r.EXPECT().Close().Return(fmt.Errorf("error"))
+		client := SetupClient(t, c, e, d)
+		_, err := client.ReadStrings(r)
+		assert.NotNil(t, err)
+	},
+	})
+	tests = append(tests, testCase{description: "ReadStrings(), with no rows, and no errors", test: func(t *testing.T) {
+		c := SetupMockConnection(t)
+		e := SetupMockEncryptor(t)
+		d := SetupMockDecryptor(t)
+		r := SetupMockRows(t)
+		r.EXPECT().Next().Return(false)
+		r.EXPECT().Err().Return(nil)
+		r.EXPECT().Close().Return(nil)
+		client := SetupClient(t, c, e, d)
+		items, err := client.ReadStrings(r)
+		assert.Nil(t, err)
+		assert.Equal(t, 0, len(items))
+	},
+	})
+	t.Parallel()
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) { test.test(t) })
+	}
+}
+
+func TestBegin(t *testing.T) {
+	type testCase struct {
+		description string
+		test        func(t *testing.T)
+	}
+
+	var tests []testCase
+
+	// Tests with shouldEncryptSet to false
+	tests = append(tests, testCase{description: "Begin(), with no errors", test: func(t *testing.T) {
+		c := SetupMockConnection(t)
+		e := SetupMockEncryptor(t)
+		d := SetupMockDecryptor(t)
+
+		sqlTx := &sql.Tx{}
+		c.EXPECT().Begin().Return(sqlTx, nil)
+		client := SetupClient(t, c, e, d)
+		txC, err := client.Begin()
+		assert.Nil(t, err)
+		assert.NotNil(t, txC)
+	},
+	})
+	tests = append(tests, testCase{description: "Begin(), with connection Begin() error", test: func(t *testing.T) {
+		c := SetupMockConnection(t)
+		e := SetupMockEncryptor(t)
+		d := SetupMockDecryptor(t)
+
+		c.EXPECT().Begin().Return(nil, fmt.Errorf("error"))
+		client := SetupClient(t, c, e, d)
+		_, err := client.Begin()
+		assert.NotNil(t, err)
+	},
+	})
+	t.Parallel()
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) { test.test(t) })
+	}
+}
+
+func TestUpsert(t *testing.T) {
+	type testCase struct {
+		description string
+		test        func(t *testing.T)
+	}
+
+	var tests []testCase
+
+	testObject := testStoreObject{Id: "something", Val: "a"}
+	// Tests with shouldEncryptSet to false
+	tests = append(tests, testCase{description: "Upsert() with no errors", test: func(t *testing.T) {
+		c := SetupMockConnection(t)
+		e := SetupMockEncryptor(t)
+		d := SetupMockDecryptor(t)
+
+		client := SetupClient(t, c, e, d)
+		txC := NewMockTXClient(gomock.NewController(t))
+		sqlStmt := &sql.Stmt{}
+		stmt := NewMockStmt(gomock.NewController(t))
+		testObjBytes := toBytes(testObject)
+		testByteValue := []byte("something")
+		e.EXPECT().Encrypt(testObjBytes).Return(testByteValue, testByteValue, testByteValue, testByteValue, nil)
+		txC.EXPECT().Stmt(sqlStmt).Return(stmt)
+		txC.EXPECT().StmtExec(stmt, "somekey", testByteValue, testByteValue, testByteValue, testByteValue).Return(nil)
+		err := client.Upsert(txC, sqlStmt, "somekey", testObject, true)
```

Comment:
Got this error:
```
pkg/cache/sql/db/client_test.go:361:24: cannot use txC (variable of type *MockTXClient) as TXClient value in argument to client.Upsert: *MockTXClient does not implement TXClient (missing method Cancel)
```

Reply:
Made some recent changes, I plan on getting back to these tests. Will push other parts first then come back to this.